### DISCUSSION TITLE
operation failed due to a quota of no more than 32500 Literals per Query

### DISCUSSION THREAD

**Discussion by rojinva** [sqlalchemy/sqlalchemy#11638] [Usage Questions]
I am using [Microsoft][ODBC Driver 17 for SQL Server][SQL Server] in a Python script to query a table. I have a simple script to execute with a list of filters. 
Example: select * from table_ABC where column_EFG in (1,2,3) group by Column_Y
In the script, I am using a variable for the list column_EFG; the list's length is more than 50000. So when I prepare the query the query will look like this: select * from table_ABC where column_EFG in (1,2,3,4 ....... , 50000) group by Column_Y
But due to the quota issue, it is throwing an error. I tried chunking the list into small and querying it multiple times, but this will not work since the query has a group-by. 
Error: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The operation failed due to a quota of no more than 32500 Literals per Query
Can you please suggest a way to overcome this?

**Comment by CaselIT** [comment]
Hi,

I think that limit comes from odbc, so you would have to look at odbc for guidance.
If the values are simple int, you may use something like `column.in_(bindparameter('a_name', the_list_here, literal_execute=True)` to instruct sqlalchemy to render the query as without bind parameters

**Reply by rojinva** [reply]
The values are a list of stings.

**Reply by CaselIT** [reply]
String should also work.
Note that's unlikely that such a query will perform great though

**Reply by rojinva** [reply]
To add more context to this issue:
This is the code what I am executing. 
from langchain_classes import SQLDatabase, LazyReflectMetadata
db = SQLDatabase.from_uri(conn_str, metadata=LazyReflectMetadata())
query = "select * from ABC where a in ('123','234')
db.run(query, include_columns=True)

**Comment by gordthompson** [comment]
You should be able to work around the limit by using OPENJSON

```python
import json

from sqlalchemy import Column, MetaData, Table, Unicode, create_engine, select, text

engine = create_engine("mssql+pyodbc://scott:tiger^5HHH@msodbcsql17_199")

t = Table("table_ABC", MetaData(), Column("column_EFG", Unicode(10)))
t.drop(engine, checkfirst=True)
t.create(engine)

list_of_values = [f"{x}" for x in range(3)]
list_as_json = json.dumps([{"x": x} for x in list_of_values])

subselect = text("SELECT x FROM OPENJSON(:json_str) WITH (x nvarchar(10) '$.x')")
qry = select(t).where(t.c.column_EFG.in_(subselect))
print(qry)
"""
SELECT "table_ABC"."column_EFG" 
FROM "table_ABC" 
WHERE "table_ABC"."column_EFG" IN (SELECT x FROM OPENJSON(:json_str) WITH (x nvarchar(10) '$.x'))
"""

engine.echo = True
with engine.begin() as conn:
    rows = conn.execute(qry, dict(json_str=list_as_json)).all()
    """
SELECT [table_ABC].[column_EFG] 
FROM [table_ABC] 
WHERE [table_ABC].[column_EFG] IN (SELECT x FROM OPENJSON(?) WITH (x nvarchar(10) '$.x'))
[generated in 0.00098s] ('[{"x": "0"}, {"x": "1"}, {"x": "2"}]',)
    """
```
